mod/wikirate/set/type_plus_right/metric/all_values.rb

```diff
@@ -0,0 +1,231 @@
+     
+    url_template = "/#{card.cardname.url_key}?item=content&offset=#{offset}"\
+                   "&limit=#{limit}&sort_order=%s&sort_by=%s"
+    %{
+      <div class='yinyang-row column-header'>
+        <div class='company-item value-item'>
+          <a class='header metric-list-header slotter' data-remote='true' 
+            href='#{sprintf(url_template, company_sort_order, 'company_name')}'>
+            Companies #{company_sort_icon}
+          </a>
+          <a class='data metric-list-header slotter' data-remote='true' 
+            href='#{sprintf(url_template, value_sort_order, 'value')}'>
+            Values #{value_sort_icon}
+          </a>
+        </div>
+      </div>
+    }
+  end
+  # compare lenght first and then normal string comparison
+  def strcmp str1, str2
+    if (length_diff = str1.length - str2.length) == 0
+      str1 <=> str2
+    else
+      length_diff
+    end
+  end
+
+  view :card_list_item do |args|
+    c = args[:item_card]
+    item_view = inclusion_defaults(c)[:view]
+    %{
+      <div class="search-result-item item-#{ item_view }">
+        #{nest(c, size: args[:size], view: item_view)}
+      </div>
+    }
+  end
+
+  view :card_list_items do |args|
+    results =
+      search_results.map do |row|
+        c = Card["#{card.cardname.left}+#{row[0]}"]
+        render :card_list_item, args.clone.merge(item_card: c)
+      end.join "\n"
+  end
+
+  view :card_list do |args|
+    paging = _optional_render :paging, args.merge({cached_query: true})
+    if search_results.empty?
+      render_no_search_results(args)
+    else
+      results = render :card_list_items, args
+      header = render :card_list_header, args
+      %{
+        #{paging}
+        #{header}
+        <div class="search-result-list">
+          #{results}
+        </div>
+        #{paging if search_results.length > 10}
+      }
+    end
+  end
+end
+
+def cached_count
+  cached_count_card.content
+end
+
+expired_cached_count_cards do |changed_card|
+  case changed_card.type_id
+  when MetricValueID
+    changed_card = changed_card.left.left.fetch(trait: :all_values)
+    changed_card.update_cached_count if changed_card
+  when WikirateCompanyID
+    metrics = Card.search type_id: MetricID, right_plus: changed_card.name
+    metrics.each do |metric|
+      changed_card = metric.fetch(trait: :all_values)
+      changed_card.update_cached_count if changed_card  
+    end
+  when MetricID
+    changed_card = changed_card.fetch(trait: :all_values)
+    changed_card.update_cached_count if changed_card  
+  end
+
+end
+
+# get all metric values
+def calculate_count
+  result = {}
+  item_cards(default_query: true).each do |card|
+    company_name = card.left.left.cardname.right
+    year = card.cardname.parts[-2]
+    value = card.content
+    unless result.has_key?company_name
+      result[company_name] = []
+    end
+    result[company_name].push({ year: year, value: value})
+  end
+  result.to_json
```

Comment:
It's clever to try to reuse this code, but we can't do this without serious renaming and/or refactoring.  We can't have a method called "calculate_count" return json with a bunch of values.  it's just too hacky.